Automattic/wp-calypso
# Checkout: create shopping-cart package and ShoppingCartProvider

#### Changes proposed in this Pull Request

This creates a new wrapper component, `ShoppingCartProvider`, which can be used along with its hook, `useShoppingCart` to provide access to the shopping cart to any component in calypso (assuming the provider is used on that route). This PR only wraps checkout in the provider, but it can be used on other routes in the future to replace the `CartData` provider.

For components which cannot use hooks, this also adds a `withShoppingCart` HOC that provides the same data.

This also creates a new package in the monorepo for the above and the rest of the shopping-cart functions and types called `@automattic/shopping-cart`.

#### Testing instructions

This is a big change for new checkout, but the shopping cart is so heavily interwoven into all aspects of checkout, if anything works, it's very likely that everything will work. The API of the shopping cart manager has _not changed_ with this PR (except for adding a way to reload the cart). This PR really just moves where the functions are defined and where they are created at runtime.

Reviewing this PR is probably most easily done commit-by-commit.

Notable things that have changed and should be verified:

- If the cart key changes, we reload the cart from scratch. This should probably have been done sooner, but previously it was unlikely that the cart key would change after checkout had loaded. Since we now have the provider being rendered at the top of a route, it's possible that something might cause the cart key to change (eg: switching to a different site without loading a new page). This is new behavior and we aren't using it currently, so I'm not very concerned.
- The shopping cart manager now requires the cart key to be set as an argument. That was mostly true before, but previously if it was falsy, the 'no-site' cart key would be used. Now the 'no-site' cart key should be provided explicitly if needed and a falsy cart key means that the cart is not ready to be loaded. This should work correctly, but it's worth testing the `no-site` behavior explicitly (this is done by the domain-only singup e2e tests).
- I don't know how to test the `isLoggedOutCart` and `isNoSiteCart` behaviors, but they are moved around by this change so they should be tested. We may be able to learn more by using the test instructions in https://github.com/Automattic/wp-calypso/pull/44206
